drivers/eeprom/eeprom_emulator.c

```diff
@@ -0,0 +1,748 @@
+	/* Device of the flash partition used to emulate the EEPROM */
+	const struct device *flash_dev;
+};
+
+struct eeprom_emu_data {
+	/* Offset in current (EEPROM) page where next change is written */
+	off_t write_offset;
+	/* Offset of the current (EEPROM) page */
+	off_t page_offset;
+	struct k_mutex lock;
+};
+
+/*
+ * basic flash read, only used with offset aligned to flash write block size
+ */
+static inline int eeprom_emu_flash_read(const struct device *dev, off_t offset,
+					uint8_t *blk, size_t len)
+{
+	const struct eeprom_emu_config *dev_config = DEV_CONFIG(dev);
+
+	return flash_read(dev_config->flash_dev, dev_config->flash_offset +
+			  offset, blk, len);
+}
+
+/*
+ * basic flash write, only used with offset aligned to flash write block size
+ */
+static inline int eeprom_emu_flash_write(const struct device *dev, off_t offset,
+				   const uint8_t *blk, size_t len)
+{
+	const struct eeprom_emu_config *dev_config = DEV_CONFIG(dev);
+
+	return flash_write(dev_config->flash_dev, dev_config->flash_offset +
+			   offset, blk, len);
+}
+
+/*
+ * basic flash erase, only used with offset aligned to flash page and len a
+ * multiple of the flash page size
+ */
+static inline int eeprom_emu_flash_erase(const struct device *dev, off_t offset,
+				   size_t len)
+{
+	const struct eeprom_emu_config *dev_config = DEV_CONFIG(dev);
+
+	return flash_erase(dev_config->flash_dev, dev_config->flash_offset +
+			   offset, len);
+}
+
+/*
+ * eeprom_emu_page_invalidate: invalidate a page by writing all zeros at the end
+ */
+static int eeprom_emu_page_invalidate(const struct device *dev, off_t offset)
+{
+	const struct eeprom_emu_config *dev_config = DEV_CONFIG(dev);
+	uint8_t buf[dev_config->flash_cbs];
+
+	LOG_DBG("Invalidating page at [0x%tx]", (ptrdiff_t)offset);
+
+	memset(buf, 0x00, sizeof(buf));
+
+	offset += (dev_config->page_size - sizeof(buf));
+	return eeprom_emu_flash_write(dev, offset, buf, sizeof(buf));
+}
+
+/*
+ * eeprom_emu_get_address: read the address from a change block
+ */
+static uint32_t eeprom_emu_get_address(const struct device *dev,
+				       const uint8_t *blk)
+{
+	const struct eeprom_emu_config *dev_config = DEV_CONFIG(dev);
+	uint32_t address = 0U;
+
+	blk += dev_config->flash_cbs/2;
+	for (int i = 0; i < sizeof(address); i++) {
+		if (2*i == dev_config->flash_cbs) {
+			break;
+		}
+
+		address += ((uint32_t)(*blk) << (8*i));
+		blk++;
+	}
+
+	return address;
+}
+
+/*
+ * eeprom_emu_set_change: create change blocks from data in blk and address
+ */
+static void eeprom_emu_set_change(const struct device *dev,
+				  const uint32_t address, const uint8_t *data,
+				  uint8_t *blk)
+{
+	const struct eeprom_emu_config *dev_config = DEV_CONFIG(dev);
+
+	for (int i = 0; i < (dev_config->flash_cbs/2); i++) {
+		(*blk++) = (*data++);
+	}
+
+	for (int i = 0; i < (dev_config->flash_cbs/2); i++) {
+		if (i < sizeof(address)) {
+			(*blk++) = (uint8_t)(((address >> (8 * i)) & 0xff));
+		} else {
+			(*blk++) = 0xff;
+		}
+
+	}
```

Comment:
Am I missing something or this loop actually overwrites what the first loop has written to the buffer pointed by `blk`?

Reply:
You are missing something, *blk goes from 0 to flash_cbs/2 - 1 in the first loop, and from flash_cbs/2 to flash_cbs - 1 in the second.